src/services/auth/utils.ts

```diff
@@ -0,0 +1,112 @@
+import { Consent } from './service/identities/consents';
+
+export type ScopeTreeLeaf = {
+  scope: string;
+  atomically_revocable: boolean;
+  children: ScopeTreeLeaf[];
+};
+/**
+ * Parses a scope string into a normalized structure (leaf) for easier comparison.
+ */
+function parseScope(s: string): ScopeTreeLeaf {
+  let parsedScope = s;
+  const revocable = parsedScope.startsWith('*');
+  if (revocable) {
+    parsedScope = parsedScope.replace(/^\*\s*/, '');
+  }
+  let children: ScopeTreeLeaf[] = [];
+  /**
+   * If there is no bracket, then there are no children and we can return the parsed scope.
+   */
+  const firstBracket = parsedScope.indexOf('[');
+  if (firstBracket === -1) {
+    return {
+      scope: parsedScope,
+      atomically_revocable: revocable,
+      children: [],
+    };
+  }
+  /**
+   * The top-level scope is everything before the first encountered bracket.
+   */
+  const topLevelScope = parsedScope.slice(0, firstBracket);
+  /**
+   * The children are everything inside the brackets.
+   */
+  children = parsedScope
+    .slice(firstBracket + 1, -1)
```

Comment:
Do we need to be concerned that this could ever end with something other than a `]`"? Maybe https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/lastIndexOf?

Reply:
The current processing expects the last character of the `parsedScope` at this point is `]`, otherwise you have malformed scope string.

...but, _maybe_ this doesn't account for space-separated deeply nested dependent scopes, I'll write a test and see.

Reply:
For sure did not work for space separated, nested dependent scopes... this should be update and the expectation that `]` is still the last character is still there. 

In general these utils are currently avoiding validation of any kind, which maybe we need eventually to avoid issues?